packages/yarnpkg-core/sources/httpUtils.ts

```diff
@@ -170,32 +178,44 @@ export enum Method {
 export type Options = {
   configuration: Configuration;
   customErrorMessage?: (err: RequestError, configuration: Configuration) => string | null;
-  headers?: {[headerName: string]: string};
+  headers?: {[headerName: string]: string | undefined};
   jsonRequest?: boolean;
   jsonResponse?: boolean;
   method?: Method;
+  wrapNetworkRequest?: (executor: () => Promise<Response>, extra: WrapNetworkRequestInfo) => Promise<() => Promise<Response>>;
 };
 
-export async function request(target: string | URL, body: Body, {configuration, headers, jsonRequest, jsonResponse, method = Method.GET}: Omit<Options, 'customErrorMessage'>) {
-  const realRequest = async () => await requestImpl(target, body, {configuration, headers, jsonRequest, jsonResponse, method});
+export async function request(target: string | URL, body: Body, {configuration, headers, jsonRequest, jsonResponse, method = Method.GET, wrapNetworkRequest}: Omit<Options, 'customErrorMessage'>) {
+  const options = {target, body, configuration, headers, jsonRequest, jsonResponse, method};
+
+  const realRequest = async () => await requestImpl(target, body, options);
+
+  const wrappedRequest = typeof wrapNetworkRequest !== `undefined`
+    ? await wrapNetworkRequest(realRequest, options)
+    : realRequest;
 
   const executor = await configuration.reduceHook(hooks => {
     return hooks.wrapNetworkRequest;
-  }, realRequest, {target, body, configuration, headers, jsonRequest, jsonResponse, method});
+  }, wrappedRequest, options);
 
   return await executor();
 }
 
-export async function get(target: string, {configuration, jsonResponse, customErrorMessage, ...rest}: Options) {
-  let entry = miscUtils.getFactoryWithDefault(cache, target, () => {
-    return prettyNetworkError(request(target, null, {configuration, ...rest}), {configuration, customErrorMessage}).then(response => {
-      cache.set(target, response.body);
-      return response.body;
-    });
-  });
-
-  if (Buffer.isBuffer(entry) === false)
-    entry = await entry;
+export async function get(target: string, {configuration, jsonResponse, customErrorMessage, wrapNetworkRequest, ...rest}: Options) {
+  const runRequest = () => prettyNetworkError(request(target, null, {configuration, wrapNetworkRequest, ...rest}), {configuration, customErrorMessage})
+    .then(response => response.body);
+
+  // We cannot cache responses when wrapNetworkRequest is used, as it can differ between calls
```

Comment:
This is not about the hook, it's about the function parameter.

The `wrapNetworkRequest` hook can indeed modify the response, but for a given URL it's kinda guaranteed that the same hooks will run even across different calls.

However, the function argument (that I'm using for caching npm metadata) can be passed different values in different calls, meaning that in one call I could pass a `wrapNetworkRequest` that returns the body `foo`, and in a different call one that returns the body `bar`, therefore breaking caching.

Thinking about it more, the entire cache in `httpUtils` is a bit unsafe because it only depends on the target, but the result could be different depending on the `configuration` (settings and hooks) / `headers` / `jsonResponse`.

I assume that it was implemented with the assumption that there won't be different option bags passed to it across different calls for a given URL in the same process, but that's in no way enforced and could lead to bugs.

In any case, I don't intend to fix that in this PR.

Reply:
> However, the function argument (that I'm using for caching npm metadata) can be passed different values in different calls, meaning that in one call I could pass a wrapNetworkRequest that returns the body foo, and in a different call one that returns the body bar, therefore breaking caching.

Yes, that's what I mean by having a pure version of the hook - one that would semantically require implementers to return the same output given the same input.

But yes, that's a follow-up, maybe not even necessary until requested.